### PR TITLE
update minimum Python to 3.10

### LINKED ISSUE #1699: use typing.TypeAlias directly when Python 3.9 is deprecated
https://github.com/mandiant/capa/blob/3cf748a135f41ac5dff86f377dae816eb529b5a0/capa/features/extractors/base_extractor.py#L15

### LINKED ISSUE #1896: Changing type annotation in simple_message_exception_handler
When capa drops support for Python 3.8, let's add in the type annotation for the exctype parameter in the simple_message_exception_handler function definition in main.py. The type annotation would be "type[BaseException]". However, the built-in type() function is not subscriptable in Python 3.8 (only Python 3.9 and above) - that is why the type annotation is not currently included.

Once we've done that, let's also take out the "# type: ignore[argument]" statement that we use in the main function to disable mypy warnings relating to the type annotation issue.

### LINKED ISSUE #1966: Remove support for Python 3.8
>  After that, it is expected that security updates (source only) will be released until 5 years after the release of 3.8 final, so until approximately October 2024. - https://peps.python.org/pep-0569/

```[tasklist]
### Tasks
- [ ] #1896
- [ ] #1912
- [x] update pyproject.toml to specify python 3.9
- [x] use dict[...] instead of Typing.Dict[...], etc.
```

### PR DESCRIPTION
**Pull request by mr-tz**
closes #1966 
closes #1896
closes #1699

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/main.py [hunk 1]
```diff
@@ -185,15 +185,11 @@ def get_default_signatures() -> List[Path]:
     return ret
 
 
-def simple_message_exception_handler(exctype, value: BaseException, traceback: TracebackType):
+def simple_message_exception_handler(
+    exctype: type[BaseException], value: BaseException, traceback: TracebackType | None
```
Comment (williballenthin) [inline capa/main.py:189]: nice syntax!

File: capa/features/extractors/base_extractor.py [hunk 2]
```diff
@@ -11,13 +11,9 @@
 import dataclasses
 from copy import copy
 from types import MethodType
-from typing import Any, Set, Dict, Tuple, Union, Iterator
+from typing import Any, Union, Iterator, TypeAlias
 from dataclasses import dataclass
 
-# TODO(williballenthin): use typing.TypeAlias directly when Python 3.9 is deprecated
-# https://github.com/mandiant/capa/issues/1699
-from typing_extensions import TypeAlias
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:19]: can we remove this from dependencies?

File: capa/features/extractors/binja/extractor.py [hunk 3]
```diff
@@ -48,12 +48,12 @@ def get_functions(self) -> Iterator[FunctionHandle]:
...
             yield FunctionHandle(address=AbsoluteVirtualAddress(f.start), inner=f)
 
-    def extract_function_features(self, fh: FunctionHandle) -> Iterator[Tuple[Feature, Address]]:
+    def extract_function_features(self, fh: FunctionHandle) -> Iterator[tuple[Feature, Address]]:
         yield from capa.features.extractors.binja.function.extract_features(fh)
 
     def get_basic_blocks(self, fh: FunctionHandle) -> Iterator[BBHandle]:
         f: binja.Function = fh.inner
-        # Set up a MLIL basic block dict look up to associate the disassembly basic block with its MLIL basic block
+        # set up a MLIL basic block dict look up to associate the disassembly basic block with its MLIL basic block
```
Comment (williballenthin) [inline capa/features/extractors/binja/extractor.py:56]: oops

File: capa/features/extractors/base_extractor.py (hunk 2 above)
Comment (mr-tz) [reply capa/features/extractors/base_extractor.py:19]: we still use `from typing_extensions import assert_never` (once) and the proto pyi files also import `typing_extensions`

**Review by williballenthin — APPROVED** [review]

File: capa/features/extractors/base_extractor.py (hunk 2 above)
Comment (mr-tz) [reply capa/features/extractors/base_extractor.py:19]: and that :)

**Pull request merged**